src/helpers/async_file_writer.rs

```diff
@@ -0,0 +1,177 @@
+                    // disarm the remove_guard.
+                    ScopeGuard::into_inner(remove_guard);
+                }
+                PkgFmt::Zip => {
+                    let mut file = tempfile()?;
+
+                    Self::read_into_file(&mut file, &mut rx)?;
+
+                    // rewind it so that we can pass it to unzip
+                    file.rewind()?;
+
+                    unzip(file, &path)?;
+                }
+                _ => extract_compressed_from_readable(
+                    ReadableRx::new(&mut rx),
+                    fmt,
+                    &path,
+                    desired_outputs.as_ref().map(|arr| &arr[..]),
+                )?,
+            }
+
+            Ok(())
+        }));
+
+        Self { handle, tx }
+    }
+
+    fn read_into_file(
+        file: &mut fs::File,
+        rx: &mut mpsc::Receiver<Content>,
+    ) -> Result<(), BinstallError> {
+        while let Some(content) = rx.blocking_recv() {
+            match content {
+                Content::Data(bytes) => file.write_all(&*bytes)?,
+                Content::Abort => {
+                    return Err(io::Error::new(io::ErrorKind::Other, "Aborted").into())
+                }
+            }
+        }
+
+        file.flush()?;
+
+        Ok(())
+    }
+
+    /// Upon error, this writer shall not be reused.
+    /// Otherwise, `Self::done` would panic.
+    async fn write(&mut self, bytes: Bytes) -> Result<(), BinstallError> {
+        if self.tx.send(Content::Data(bytes)).await.is_err() {
+            // task failed
+            Err(Self::wait(&mut self.handle).await.expect_err(
+                "Implementation bug: write task finished successfully before all writes are done",
+            ))
+        } else {
+            Ok(())
+        }
+    }
+
+    async fn done(mut self) -> Result<(), BinstallError> {
+        // Drop tx as soon as possible so that the task would wrap up what it
+        // was doing and flush out all the pending data.
+        drop(self.tx);
+
+        Self::wait(&mut self.handle).await
+    }
+
+    async fn wait(
+        handle: &mut AutoAbortJoinHandle<Result<(), BinstallError>>,
+    ) -> Result<(), BinstallError> {
+        match handle.await {
+            Ok(res) => res,
+            Err(join_err) => Err(io::Error::new(io::ErrorKind::Other, join_err).into()),
+        }
+    }
+
+    fn abort(self) {
+        let tx = self.tx;
+        // If Self::write fail, then the task is already tear down,
+        // tx closed and no need to abort.
+        if !tx.is_closed() {
+            // Use send here because blocking_send would panic if used
+            // in async threads.
+            tokio::spawn(async move {
+                tx.send(Content::Abort).await.ok();
+            });
+        }
+    }
+}
+
+/// AsyncFileWriter removes the file if `done` isn't called.
+#[derive(Debug)]
+pub struct AsyncFileWriter(ScopeGuard<AsyncFileWriterInner, fn(AsyncFileWriterInner), Always>);
```

Comment:
It would be useful to maybe rename this or add some more docs, because it's confusing as to what it does. There's unzipping? There's reading? There's what appears to be an async byte stream implemented over a channel?

Reply:
Honestly, I'm not good at writing doc alone.

So perhaps you can keep asking me question about internals of `AsyncFileWriter` and I will answer it here while adding it to the doc.

Reply:
Well, I understand well enough how it works, I've read it, the issue here if treating it as opaque, looking at it from the outside, like me in 6 months, or anyone who wants to hack on this in the future, it's confusing.

Let me walk through a thought process, here:

- Ok, this is something called an async file writer. So, I'm expecting it writes a file, or maybe some files, asyncly. Ok.
- So, it takes a path, alright, that must be the path of the file it's writing,
- and a format, which looks like an archive format, so maybe it writes that file in this format,
- and a... desired outputs array? what?
- Uh, let's look at the documentation...
- it passes the `Bytes` you give it (wait but nowhere in the signature do you pass bytes, oh it's via the `write` method? uhh)
- to another thread via an mpsc (weird but okay, sure, why?)
- and decompresses and unpacks if needed.
- okay, so actually, it _reads_ from a possibly-compressed file, so, okay, the path must be to the file to read
- wait but you _give_ it bytes?
- so, it reads from a `Bytes` that you give it... and writes files to... a path? fine i guess
- and wait this cancellation bit... it "removes the file if done is not called"?
- which file? one of the ones it wrote? which one?
- is it a typo? is it all of the files it wrote?
- it can't be the possibly-compressed file, right, because I'm just giving it as bytes
- also, why am I giving it as bytes? shouldn't I just give it whole if I have it?
- ohhh, this is meant to have, like a stream of data coming in, right?
- weird, but okay

So, this is the outlook I'm looking to clarify here by docs and/or names.

(But hold on, there's a part 2!)

Reply:
I renamed `AsyncFileWriter` to `AsyncExtracter`, it should be more straight forward of what it is doing.

Reply:
I've added more doc, added part of the PR description to it and rename `AsyncExtracter::write` to `feed`.

Reply:
Part 2 is, okay, so what we have is a streaming I/O processor that consumes an archive (and also sometimes just a plain bin) and writes some specified files within it to disk. Except if the archive is zip, in which case it will write everything that's contained.

So, that's a bit complicated. Can we simplify that? Can we make:
- a streaming I/O processor
- that consumes an archive
- and writes the specified contained files to disk.

An ArchiveStreamExtractor, let's say.

And then we'll specify that you can't call it with a plain bin. In fact we could make our own `ArchiveFormat` enum that's split from `PkgFmt` because `PkgFmt` is about the binstall interface and metadata and URLs and templates, and this new `ArchiveFormat` is actually _just_ about how to unpack this archive stream we have here. `ArchiveFormat` will _not_ include `Bin`, that way the caller is forced to do something else for that, like directly write it to a file.

Also (unless I've misunderstood the caveats here), we'll uphold our promise of writing only specified files for `Zip` by postprocessing and deleting files that don't match. That way it's always clear and obvious what the output will be, even though it's a bit hacky on the inside at the moment (and who knows, maybe later we can improve and do partial extraction with `Zip`).

So, that's a good place to start.

And then there's part 3, which is some questions about the internals:
- Instead of the `write` and `done` methods, could we take an `impl Stream<Item = Result<Bytes>>` and iterate it ourselves then finalise (call done)?
- Instead of a new ReadableRx type, could we write the bytes into a [`DuplexStream`](https://docs.rs/tokio/latest/tokio/io/struct.DuplexStream.html)? That seems to have been designed for precisely this purpose.

Reply:
And then we don't really _need_ to have better docs, because our public API for this will look like:

```rust
pub async fn extract_archive_stream(
    stream: impl Stream<Item = Result<Bytes>>,
    output_dir: impl AsRef<Path>,
    file_filter: Option<Vec<Cow<Path>>>,
) -> Result<()>
```

Reply:
> And then we'll specify that you can't call it with a plain bin.

I actually consider an advantage.

Without `AsyncExtracter`, you would do potentially blocking file writing on the async thread, which means blocking the tokio runtime (which is a bad practice).

> Also (unless I've misunderstood the caveats here), we'll uphold our promise of writing only specified files for Zip by postprocessing and deleting files that don't match. That way it's always clear and obvious what the output will be, even though it's a bit hacky on the inside at the moment (and who knows, maybe later we can improve and do partial extraction with Zip).

I don't think that part is necessary.
The `desired_outputs` is a best-effort: If you could filter the outputs, then great, it saves I/O and disk usage.
If not, then simply extract the whole content is ok.

> Instead of the write and done methods, could we take an impl Stream<Item = Result<Bytes>> and iterate it ourselves then finalise (call done)?

Yeah, definitely could do.

> Instead of a new ReadableRx type, could we write the bytes into a [DuplexStream](https://docs.rs/tokio/latest/tokio/io/struct.DuplexStream.html)? That seems to have been designed for precisely this purpose.

No.

Currently, we read directly from the `Bytes` received and it requires one copy.
Adding `DuplexStream` adds another buffer and another copy.

Reply:
> And then we don't really _need_ to have better docs, because our public API for this will look like:
> 
> ```rust
> pub async fn extract_archive_stream(
>     stream: impl Stream<Item = Result<Bytes>>,
>     format: ArchiveFormat,
>     output_dir: impl AsRef<Path>,
>     file_filter: Option<Vec<Cow<Path>>>,
> ) -> Result<()>
> ```

Done!